bin/specextract

```diff
@@ -79,381 +81,449 @@ v4 = make_verbose_level(toolname, 4)
-#  matching (TSTART-sorted) input asol files to each input source file.
-#  For now, 'stack' should be of the form 'stack=stk_build(files)'.
-#
-###########################################################################
-
-def get_keyvals(stack, keyname):
 
+def create_hrc_resp(specfile,rmf_file,refcoord,full_outroot,
+                    asp_param,clobber,verbose,
+                    mskfile,skyx,skyy,instrument,chip_id,infostr):
     """
-    return keys
+    Generate HRC ARF with mkarf and copy RMF from CalDB
     """
+    
+    # use the 'caldb4' module to query for the appropriate,
+    # latest HRC RMF in CALDB;  can use calquiz instead as well,
+    # output is a string rather than list
+
+    if rmf_file.upper() == "CALDB":
+        # cf. https://cxc.cfa.harvard.edu/cal/Hrc/detailed_info.html#rmf for HRC RMF information
+ 
+        rmf = caldb4.Caldb(infile=specfile,product="MATRIX").search
+
+        if len(rmf) == 0:
+            raise IOError(f"{specfile} is an invalid HRC spectral file, as non-SAMP responses are no longer supported by the CALDB.  Please reprocess the dataset or re-download the observation from the archive.")
+
+        if len(rmf) > 1:
+            # this should not happen
+            raise IOError("Multiple HRC RMFs returned by CALDB: {}".format(",".join(rmf)))
+
+        rmf = rmf[0]
+        rmf = rmf[:rmf.find("[")]
+
+        # copy RMF
+        rmffile = f"{full_outroot}.rmf"
+        shutil.copyfile(rmf,rmffile)
+
+    else:
+        # enable RMFFILE parameter for HRC data to support HRC Cal group
+        v1("Warning: 'rmffile' parameter for HRC observations is meant for Calibration Group and expert usage!")
+
+        rmffile = rmf_file
+
+    # establish detsubsys       
+    if instrument == "HRC":
+        if chip_id == "0":
+            detname = "HRC-I"
+        else:
+            detname = f"HRC-S{chip_id}"
+
+    with new_pfiles_environment(ardlib=True,copyuser=False):
+        # ARF output file
+        v1(infostr)
 
-    keys = []
+        arffile = f"{full_outroot}.arf"
+
+        mkarf.punlearn()
+
+        mkarf.detsubsys = detname
+        mkarf.outfile = arffile
+        mkarf.asphistfile = asp_param
+        mkarf.sourcepixelx = skyx
+        mkarf.sourcepixely = skyy
+        mkarf.grating = fileio.get_keys_from_file(specfile)["GRATING"]
+        mkarf.obsfile = specfile
+        mkarf.maskfile = mskfile
+        mkarf.verbose = str(verbose)
+        mkarf.clobber = clobber
 
-    for fn in stack:
         try:
-            kval = fileio.get_keys_from_file(fn)[keyname]
-        except KeyError:
-            v1(f"WARNING: The {keyname} header keyword is missing from {fn}")
-            kval = None
-			
-        keys.append(kval)
-	
-    return keys
+            mkarf.engrid = f"grid({rmffile}[SPECRESP MATRIX][cols ENERG_LO,ENERG_HI])"
+            mkarf()
 
+        except OSError:        
+            mkarf.engrid = f"grid({rmffile}[MATRIX][cols ENERG_LO,ENERG_HI])"
```

Comment:
Can we not identify from the `rmffile` argument whether to use [MATRIX] or [SPECRESP MATRIX] rather than have this try/except pattern? I'm sure we have some code that list the block names in a file.

Reply:
not directly from what I can tell, but using some variation of:


```
    blnames = [bl[0].lower() for bl in ciao_contrib.cxcdm_wrapper.get_info_from_file(rmffile)]

    if "specresp matrix" in blnames:
        blname = "SPECRESP MATRIX"
    else:
        blname = "MATRIX"

```

will do the trick to avoid try/except.